### PR TITLE
feat(planning_validator): add planning validator package

### PR DESCRIPTION
**Pull request by TakaHoribe**
## Description

For the Issue: https://github.com/autowarefoundation/autoware.universe/issues/2605

This PR adds a new `planning_validator` pkg.

**motivation** 
- Invalid trajectory (e.g. contains infinite values, shape is broken, etc) should be detected in the planning module and handled in an appropriate way.
- the `planning_validator` will be put at the last stage in the planning module to check the validity of the planning output.

**Design**
- check some metrics
- when an invalid part is found, the `plannig_validator` will do some of the following list based on the parameter:
    - publish a Diagnostic msg
    - publish last validated trajectory / publish the invalid trajectory as it is / publish nothing

## Related links

autoware_launch: https://github.com/tier4/autoware_launch/pull/675
(for TIER IV internal tracker: https://tier4.atlassian.net/browse/T4PB-18620)

## Tests performed

Now some metrics can be seen here.

https://user-images.githubusercontent.com/21360593/192083444-fba23943-b17e-432e-996b-79243bb03dde.mp4

------

**Invalid case check:** 
it is confirmed this PR works for the invalid trajectory. This is the case when the trajectory shape is broken when generating lane change trajectory. (For TIER IV INTERNAL, it can be found in [this ticket.](https://tier4.atlassian.net/browse/T4PB-23754) )

https://user-images.githubusercontent.com/21360593/210230742-772729bf-db20-4ae9-abf6-f9d35a8ed51f.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by taikitanaka3** [comment]
@TakaHoribe 
TODO in description is done by planning_error_monitor or maybe by planning evaluator.

**Comment by TakaHoribe** [comment]
@taikitanaka3 right. This package is based on the planning_error_monitor and his all features are (will be) inherited to this package.

**Comment by TakaHoribe** [comment]
@TakaHoribe

**Comment by TakaHoribe** [comment]
@TakaHoribe

**Comment by xmfcx** [comment]
In addition, the validation of a generated path is also be done by iteratively simulating the control stack on the given trajectory and seeing how well it can follow it by calculating the error.

> invalid trajectory (e.g. contains infinite values, shape is broken, etc) should be detected in the planning module

Is this package going to check for just these?

Maybe it'd be better if we can list the features that this node can validate.

**Comment by taikitanaka3** [comment]
@TakaHoribe 
In my opinion, this validator package should use trajectory from obstacle cruise and output trajectory for smoother, because smootheris necessary to calculate desired motion (eg. stop, cruise, accelerate) using previous trajectory.

**Comment by TakaHoribe** [comment]
> it'd be better if we can list the features that this node can validate.

@xmfcx Thank you for the good advice. The current implementation checks the following items:

Ported from `planning_error_monitor`
- invalid field (e.g. Inf, Nan) 
- trajectory points interval (invalid if the distance of trajectory point is too large, like 100m)
- curvature (invalid if the trajectory has too sharp turns that is not feasible for the given vehicle kinematics)

Implemented in this PR

- relative angle in trajectory points (invalid if the yaw angle changes too fast in the sequence of trajectory points)
- lateral acceleration (invalid if the expected lateral acceleration/deceleration is too large)
- longitudinal acceleration/deceleration (invalid if the acceleration/deceleration is too large)
- steering angle (expected value estimated from curvature)
- steering angle rate (expected value estimated from curvature)
- velocity deviation (invalid if the planning velocity is too far from the ego velocity)
- distance deviation (invalid if the ego is too far from the trajectory)

I'll list them in the readme as well. Any comments are welcome.

**Comment by TakaHoribe** [comment]
> In my opinion, this validator package should use trajectory from obstacle cruise and output trajectory for smoother, because smootheris necessary to calculate desired motion (eg. stop, cruise, accelerate) using previous trajectory.

@taikitanaka3 Sorry I don't get your point. Does your idea solve the original motivation: "we don't want to send the invalid trajectory to the controller"?

**Comment by taikitanaka3** [comment]
@TakaHoribe 
partly yes, but my concern is Is it possible to drive using previous trajectory without smoother all the time.

**Comment by yukkysaito** [comment]
Can you share CPU utilization of this package?

**Comment by TakaHoribe** [comment]
@yukkysaito No heavy computational logic is used now. It is just about 2% in planning_simulator on my 4-year-old desktop PC (`Intel® Core™ i7-8700K CPU @ 3.70GHz × 12`). 
image

File: planning/planning_validator/include/planning_validator/utils.hpp
Comment (yukkysaito) [inline planning/planning_validator/include/planning_validator/utils.hpp:51]: It would be better to pass it as a smart pointer or argument since memory copying may occur.

Comment (TakaHoribe) [reply planning/planning_validator/include/planning_validator/utils.hpp:51]: Thank you for the comment. However, using smart pointer or reference argument just for this purpose is not preferred by me due to the readability. So many discussions for NRVO and no clear answer yet. I decided to leave it to optimization, and believe that is better in almost all cases recently.

Comment (yukkysaito) [reply planning/planning_validator/include/planning_validator/utils.hpp:51]: I understand how you feel.
It would be nice if NVRO was guaranteed, but currently it is no.
Because of those limitations, many codes employ argument passing by reference.

http://opencv.jp/opencv-2.1/cpp/geometric_image_transformations.html
http://docs.ros.org/en/indigo/api/tf/html/c++/classtf_1_1Transformer.html

Comment (takam5f2) [reply planning/planning_validator/include/planning_validator/utils.hpp:51]: @yukkysaito @TakaHoribe 
(Sorry for long message, but this is translated by DeepL. )

In my opinion, readability should be prioritized at first because the overhead gives too small influence to Autoware's functionality.

I tried measuring elapsed time on both "returning an object" and "returning a pointer" briefly .  When a function returns a Trajectory object which has 1000 trajectory points, about 10~20 "micro"seconds cost.  
On the other hand, returning pointer costs  about 50 ~ 200 "nano"seconds.   
I found that "returning the object " sometimes costs 100 times as much time as "returning the pointer".  
As I'm not sure how large data `Trajectory` object has, it might not be correct. 

There is exactly a large gap between returning the object and returning the pointer.  
I think that this impact is too small to influence on Autwoare's functionality if `resampleTrajectory()` is called at 10 Hz. If it is called frequently like at 1000 Hz, we may need to take care of the overhead of returning the object.

My opinion came from my bias, but I think other opinions should be taken care of.

Comment (TakaHoribe) [reply planning/planning_validator/include/planning_validator/utils.hpp:51]: @takam5f2 @yukkysaito Thank you for your comment. I've tested how the computational time is improved. Changing all implementation to reference argument improves 0.006 [ms] on average for the whole process in 500 times operation with 166 points trajectory (trajectory usually has about 100~200 points). Even with 1500 points trajectory, the improvement was 0.012 [ms]. Readability should not be sacrificed for this small improvement.

>  if NRVO was guaranteed, but currently it is no.

I know the background. Even if it is not guaranteed, it is optimized anyway. It should be ok.

> Also, allowing this would allow the point cloud to be the return value.

I totally doubt that. It should depend on the implementation. We definitely shouldn't make a copy just to change a few fields in a large object.

Well, we shouldn't take time in this discussion in this minor thread. Actually, there are already many implementations using "return large object" in the autoware. If you really care about this point, please make a discussion thread and put the decision into the coding guideline.

Here I'll take the new implementation to skip the discussion.

Comment (TakaHoribe) [reply planning/planning_validator/include/planning_validator/utils.hpp]: @takam5f2 @yukkysaito Sorry I misunderstood. ["Using reference argument](https://github.com/autowarefoundation/autoware.universe/pull/1947/commits/e81c91bafc0e61eaa9b6fa63feabba96205470ff)" is 0.006 [ms] SLOWER than "[returning object](https://github.com/autowarefoundation/autoware.universe/pull/1947/commits/f36c7820ba47ccd3fbcd614e0aca0c414750b9cf)".

Comment (yukkysaito) [reply planning/planning_validator/include/planning_validator/utils.hpp]: Thank you. Memory access is more of a concern than execution time, as memory access is currently tight on TIER IV computers.

>I know the background. Even if it is not guaranteed, it is optimized anyway. It should be ok.

Sorry. Please tell me what meaning? :pray: 
The optimization is running in the current environment, but if we change the code slightly or environment, it may not be optimized.

Comment (yukkysaito) [reply planning/planning_validator/include/planning_validator/utils.hpp]: > If you really care about this point, please make a discussion thread and put the decision into the coding guideline.

Yes. Let's talk separately.

**Comment by TakaHoribe** [comment]
@xmfcx Here is the readme, FYI.
https://github.com/tier4/autoware.universe/tree/feature/planning-validator/planning/planning_validator

> I'll list them in the readme as well. Any comments are welcome.

File: planning/planning_validator/README.md
Comment (tkimura4) [inline planning/planning_validator/README.md:59]: threholds -> thresholds

Comment (TakaHoribe) [reply planning/planning_validator/README.md]: done. https://github.com/autowarefoundation/autoware.universe/pull/1947/commits/47518bf0ec54160c682e6ba5e2da10e06ea78d33

thanks

**Review by tkimura4 — APPROVED** [review]
LGTM

**Review by tkimura4 — APPROVED** [review]

File: planning/planning_validator/include/planning_validator/planning_validator.hpp
Comment (takayuki5168) [inline planning/planning_validator/include/planning_validator/planning_validator.hpp]: typo: PUBLISH_AS_IT_**IS**?

File: planning/planning_validator/src/debug_marker.cpp
Comment (takayuki5168) [inline planning/planning_validator/src/debug_marker.cpp:62]: I feel you do not need to modify in this PR, but you can use the following function used everywhere in the planning packages.
https://github.com/autowarefoundation/autoware.universe/blob/5f9d8478c97d8991cd5dd1ec85c9b5a50d582654/common/tier4_autoware_utils/include/tier4_autoware_utils/ros/marker_helper.hpp#L67-L70

File: planning/planning_validator/src/planning_validator.cpp
Comment (takayuki5168) [inline planning/planning_validator/src/planning_validator.cpp:322]: (I feel you do not need to modify in this PR)

Just comment. I prefer
```cpp
boost::optional<size_t> calcMaxCurvatureIndex(Trajectory);
```

```cpp
  const auto idx = calcMaxCurvatureIndex(trajectory);
  if (!idx) {
    return true;
  }

  validation_status_.max_curvature = idx.get();
  if (max_curvature > validation_params_.curvature_threshold) {
    const auto & p = trajectory.points;
    debug_pose_publisher_->pushPoseMarker(p.at(i - 1), "trajectory_curvature");
    debug_pose_publisher_->pushPoseMarker(p.at(i), "trajectory_curvature");
    debug_pose_publisher_->pushPoseMarker(p.at(i + 1), "trajectory_curvature");
    return false;
  }
  return true;
```

since I disagree with returning `0` as not-enough points as follows.
```cpp
std::pair<double, size_t> calcMaxCurvature(const Trajectory & trajectory)
{
  if (trajectory.points.size() < 3) {
    return {0.0, 0};
  }
```

File: planning/planning_validator/include/planning_validator/planning_validator.hpp
Comment (TakaHoribe) [reply planning/planning_validator/include/planning_validator/planning_validator.hpp]: it was typo. THanks. Now fixed to as_it_is

File: planning/planning_validator/src/debug_marker.cpp
Comment (TakaHoribe) [reply planning/planning_validator/src/debug_marker.cpp:62]: Thanks! I'll take it

File: planning/planning_validator/src/planning_validator.cpp
Comment (TakaHoribe) [reply planning/planning_validator/src/planning_validator.cpp:322]: The 0 curvature for trajectory sounds ok to me, it is based on the definition.
But yes, as you say, returning 0 index for the empty trajectory doesn't make sense.

File: planning/planning_validator/src/debug_marker.cpp
Comment (TakaHoribe) [reply planning/planning_validator/src/debug_marker.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/1947/commits/edd358017bff62f922c1930a695201117060918e

File: planning/planning_validator/include/planning_validator/utils.hpp
Comment (kenji-miyake) [reply planning/planning_validator/include/planning_validator/utils.hpp]: Commented here: https://github.com/orgs/autowarefoundation/discussions/3209#discussioncomment-4723928

File: planning/planning_validator/src/planning_validator.cpp
Comment (TakaHoribe) [reply planning/planning_validator/src/planning_validator.cpp:322]: @takayuki5168 Let me think later, maybe updated in another PR. Please keep this comment open until that time.

**Comment by TakaHoribe** [comment]
scenario test works well: [822/825](https://evaluation.tier4.jp/evaluation/reports/e62da16d-9081-5e96-9f02-820e9ed947cd?project_id=prd_jt) -> [821/825](https://evaluation.tier4.jp/evaluation/reports/82985422-b939-58f0-804e-890138df31cc?project_id=prd_jt)

**Review by purewater0901 — APPROVED** [review]
LGTM

**Comment by TakaHoribe** [comment]
@ito-san Would you take a review as a maintainer of the `system_error_monitor`? Your review is required.

**Review by ito-san — APPROVED** [review]
LGTM

File: planning/planning_validator/CMakeLists.txt
Comment (wep21) [inline planning/planning_validator/CMakeLists.txt:46]:
```suggestion
# to use a message defined in the same package
rosidl_get_typesupport_target(
    cpp_typesupport_target ${PROJECT_NAME} "rosidl_typesupport_cpp"
)
target_link_libraries(planning_validator_component "${cpp_typesupport_target}")
```

Comment (TakaHoribe) [reply planning/planning_validator/CMakeLists.txt]: Thank you!

**Review by yukkysaito — APPROVED** [review]

File: planning/planning_validator/test/src/test_planning_validator_helper.cpp
Comment (wep21) [inline planning/planning_validator/test/src/test_planning_validator_helper.cpp:78]: typo?

Comment (TakaHoribe) [reply planning/planning_validator/test/src/test_planning_validator_helper.cpp:78]: wow

Comment (TakaHoribe) [reply planning/planning_validator/test/src/test_planning_validator_helper.cpp:78]: thank you
https://github.com/autowarefoundation/autoware.universe/pull/2707

**Pull request merged**
